MarkBind/vue-strap
# Set up jest snapshot testing for vue-strap

**What is the purpose of this pull request? (put an "X" next to item)**

•  [ ] Documentation update
•  [ ] Bug fix
•  [ ] New feature
•  [ ] Enhancement to an existing feature
• [x] Other, please explain: Added Testing


Fixes https://github.com/MarkBind/markbind/issues/409



**What is the rationale for this request?**

There was no testing framework set up for vue-strap at all.

Having a testing framework will be useful to check that we did not break anything when adding new changes, especially if Vue generates the correct DOM tree given a component with certain attributes.

**What changes did you make? (Give an overview)**

The commits in the PR are split up into ~5 4~ 3 commits for easy reviewing
- ~Added lint fix script and fix eslint errors as in #144~ 
  - ~This was necessary because I faced many CRLF/LF linting errors that needed quick fixing via an `npm run lintfix` command.~ 
- Setup and configured Jest for the project (with coverage collection on)
  - Referenced heavily from https://vue-test-utils.vuejs.org/guides/testing-single-file-components-with-jest.html
  - Created `babel.config.js` and removed `babel-loader.options` in `webpack.common.js` [since the webpack bundler will find the config file](https://babeljs.io/docs/en/config-files#project-wide-configuration). Generated a bundle and found no regressions when using the bundle and running `markbind serve` on `markbind/docs`. **(A)**
- Updated package lock file for the above two changes
- Added a test for TipBox **(B)**
  - Snapshot testing seemed sufficient. Except for tests like "`light` style take priority over `seamless`", which can be an equality check on two different mounted vue instances. **Corresponding update snapshot script was made**


**Additional note**
- Generated coverage **(is now gitignore-ed)** 
  - The coverage shows for all, including untested, Vue [SFCs](https://vuejs.org/v2/guide/single-file-components.html) we have. Something to work towards as you will see an increase coverage for every test added and run. **(Will not show up in commits/PR to reduce clutter, but feel free to see it in your own local repos)**

**Provide some example code that this change will affect:**

\-

**Is there anything you'd like reviewers to focus on?**

1. Confirm if point **(A)** is correct.
1. **(B)** Look at the tests I made for TipBox.vue and if there any improvements, perhaps in terms of 'test naming'. 
   1. I tried to follow descriptive lines that describes what the vue component _should_ be doing. I did consider trying to test whether a particular class name is present in the generated dom, but that seemed very overkill. 
    1. The idea behind snapshot testing is having a snapshot that has been checked to be correct. After which subsequent tests run are supposed to generate identical snapshots, unless a new feature is made, then which a new snapshot is updated and the updated snapshot committed. 

~One nit you may notice is that if you run tests when everything is unchanged, coverage will show as diffs because of updated timestamps in the coverage files. Personally I just discard all changes under the coverage directory unless I really updated something in it~

This PR sets the groundwork for future testing and I do not really intend to add all the tests in the vue-strap project in a single PR. Having this PR through first, enables multiple people to add tests concurrently.

**Testing instructions:**

1. Checkout this branch. And `npm install`
1. `npm run test` to run tests. You will be able to see coverage report too.
![image](https://user-images.githubusercontent.com/27397021/80473373-a117e500-8978-11ea-8561-575cc0deb24f.png)
   1. A detailed view of the report per-file per-line can be seen in in any html file located in `/coverage` 
1. You may also play around with the snapshots.

**Commit Message**

Set up tests for vue-strap

There was no testing framework set up for vue-strap at all.
Let us setup up and configure jest for this project
And add a test file for Boxes as a proof of concept
